modules/calib3d/src/ptsetreg.cpp

```diff
@@ -501,6 +501,242 @@ class Affine3DEstimatorCallback : public PointSetRegistrator::Callback
     }
 };
 
+class Affine2DEstimatorCallback : public PointSetRegistrator::Callback
+{
+public:
+    int runKernel( InputArray _m1, InputArray _m2, OutputArray _model ) const
+    {
+        Mat m1 = _m1.getMat(), m2 = _m2.getMat();
+        const Point2f* from = m1.ptr<Point2f>();
+        const Point2f* to   = m2.ptr<Point2f>();
+
+        // we need 3 points to estimate affine transform
+        const int N = 6;
+        double buf[N*N + N + N];
+        Mat A(N, N, CV_64F, &buf[0]);
+        Mat B(N, 1, CV_64F, &buf[0] + N*N);
+        Mat X(N, 1, CV_64F, &buf[0] + N*N + N);
+        double* Adata = A.ptr<double>();
+        double* Bdata = B.ptr<double>();
+        A = Scalar::all(0);
+
+        for( int i = 0; i < (N/2); i++ )
+        {
+            Bdata[i*2] = to[i].x;
+            Bdata[i*2+1] = to[i].y;
+
+            double *aptr = Adata + i*2*N;
+            for(int k = 0; k < 2; ++k)
+            {
+                aptr[0] = from[i].x;
+                aptr[1] = from[i].y;
+                aptr[2] = 1.0;
+                aptr += N+3;
+            }
+        }
+
+        solve(A, B, X, DECOMP_SVD);
+        X.reshape(1, 2).copyTo(_model);
+        return 1;
+    }
+
+    void computeError( InputArray _m1, InputArray _m2, InputArray _model, OutputArray _err ) const
+    {
+        Mat m1 = _m1.getMat(), m2 = _m2.getMat(), model = _model.getMat();
+        const Point2f* from = m1.ptr<Point2f>();
+        const Point2f* to   = m2.ptr<Point2f>();
+        const double* F = model.ptr<double>();
+
+        int count = m1.checkVector(2);
+        CV_Assert( count > 0 );
+
+        _err.create(count, 1, CV_32F);
+        Mat err = _err.getMat();
+        float* errptr = err.ptr<float>();
+
+        for(int i = 0; i < count; i++ )
+        {
+            const Point2f& f = from[i];
+            const Point2f& t = to[i];
+
+            double a = F[0]*f.x + F[1]*f.y + F[2] - t.x;
+            double b = F[3]*f.x + F[4]*f.y + F[5] - t.y;
+
+            errptr[i] = (float)(a*a + b*b);
+        }
+    }
+
+    bool checkSubset( InputArray _ms1, InputArray _ms2, int count ) const
+    {
+        Mat ms1 = _ms1.getMat(), ms2 = _ms2.getMat();
+        // check colinearity and also check that points are too close
+        return !(haveCollinearPoints(ms1, count) || haveCollinearPoints(ms2, count));
```

Comment:
I think you should only check _ms1 for collinear points and not _ms2.  The estimator is only unstable when _ms1 has collinear points -- it doesn't matter if _ms2 does or not.  Also, consider the case where the true affine transform is degenerate (i.e. maps all input points to a line).  This case may not be common in CV, but it is valid and there is no reason not to produce an accurate estimate of the transform in this case.  By checking _ms2, this case won't be able to be estimated and will error out instead.  Plus it's faster to only check _ms1.